File: node/ast.js
class ASTNode {
  constructor(type) {
    this.type = type;
  }
}

class NumberNode extends ASTNode {
  constructor(value) {
    super('NUMBER');
    this.value = value;
  }
}

class BinaryOpNode extends ASTNode {
  constructor(left, operator, right) {
    super('BINARY_OP');
    this.left = left;
    this.operator = operator;
    this.right = right;
  }
}

class UnaryOpNode extends ASTNode {
  constructor(operator, operand) {
    super('UNARY_OP');
    this.operator = operator;
    this.operand = operand;
  }
}

class RenderNode extends ASTNode {
  constructor(expression) {
    super('RENDER');
    this.expression = expression;
  }
}

class AssignmentNode extends ASTNode {
  constructor(name, value) {
    super('ASSIGNMENT');
    this.name = name;
    this.value = value;
  }
}

class VariableNode extends ASTNode {
  constructor(name) {
    super('VARIABLE');
    this.name = name;
  }
}

module.exports = { ASTNode, NumberNode, BinaryOpNode, UnaryOpNode, RenderNode, AssignmentNode, VariableNode };

File: ruby/ast.rb
# AST node classes for the calculator language
module AST
  class Node
    def accept(visitor)
      visitor.visit(self)
    end
  end

  class IntegerNode < Node
    attr_reader :value

    def initialize(value)
      @value = value
    end
  end

  class BinaryOpNode < Node
    attr_reader :left, :operator, :right

    def initialize(left, operator, right)
      @left = left
      @operator = operator
      @right = right
    end
  end

  class UnaryOpNode < Node
    attr_reader :operator, :operand

    def initialize(operator, operand)
      @operator = operator
      @operand = operand
    end
  end

  class RenderNode < Node
    attr_reader :expression

    def initialize(expression)
      @expression = expression
    end
  end
end